src/Orleans/Streams/Internal/StreamDirectory.cs

```diff
@@ -20,7 +20,14 @@ internal StreamDirectory()
 
         internal IAsyncStream<T> GetOrAddStream<T>(StreamId streamId, Func<IAsyncStream<T>> streamCreator)
         {
-            return allStreams.GetOrAdd(streamId, _ => streamCreator()) as IAsyncStream<T>;
+            var stream = allStreams.GetOrAdd(streamId, _ => streamCreator());
+            var streamOfT = stream as IAsyncStream<T>;
+            if (streamOfT == null)
+            {
+                throw new Runtime.OrleansException($"Stream type mismatch. A stream can only support a single type of data. The generic type of the stream requested does not match the previously requested type {stream.GetType()}.");
+            }
```

Comment:
Looks good, but maybe:
The generic type of the stream requested ({typeof(T)}), does not match the previously requested type ({stream.GetType().GetGenericArguments().FirstOrDefault()}).